### PR TITLE
wgsl: Add utility to help check for ambiguous overflows

### LINKED ISSUE #3623: `webgpu:shader,validation,expression,call,builtin,distance:values:type="f16";*` test failures
https://github.com/gpuweb/cts/pull/3585 introduced `distance()` validation tests.

These tests fail for `f16` with the error:

```
  - EXCEPTION: Error: Unexpected validation error occurred: Tint program failure: :4:18 error: '-65504.0 - 0.062469482421875' cannot be represented as 'f16'
    var<private> v = distance(f16(o0), f16(o1));
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    :4:18 note: when calculating distance
    var<private> v = distance(f16(o0), f16(o1));
```

The finite representable range for `f16` is `[-65504.0 .. +65504.0]`, so this looks like a valid failure (Tint is correct, CTS is wrong).

The `distance()` CTS attempts to detect the non-representable intermediate with:

https://github.com/gpuweb/cts/blob/afef918253ae6ba28bee20dfbf5b4e980103949f/src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts#L65-L74

For `Type.f16`, `quantizeFn()` calls `quantizeToF16()`, which in turn calls `hfround()`, which is implemented as:
https://github.com/gpuweb/cts/blob/afef918253ae6ba28bee20dfbf5b4e980103949f/src/external/petamoriken/float16/float16.js#L1219-L1226
https://github.com/gpuweb/cts/blob/afef918253ae6ba28bee20dfbf5b4e980103949f/src/external/petamoriken/float16/float16.js#L344-L349

This does not appear to convert values outside of the finite range to infinity.

Interestingly, the `petamoriken/float16` library on GitHub appears substantially different, and does treat these as infinities:

https://github.com/petamoriken/float16/blob/60551c22d2eba68c829e04da8956a917786e7a92/src/_util/converter.mjs#L32-L63

@zoddicus - Should we try to update our version of `petamoriken/float16` ?

### PR DESCRIPTION
**Pull request by toji**
Fixes #3623

Primary thing to pay attention to is the new `ConstantOrOverrideValueChecker` utility class (rolls off the tongue, huh?) This is performing the same quantization step as before but now has the opportunity to force a subcase to be skipped if the checked value is found to be in a range where the overflow rounding behavior is ambiguous.

If we think this is a good direction then I'll go back and update a bunch of the tests I've already completed to use the new utility class for consistency/completeness.

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [x] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts [hunk 1]
```diff
@@ -312,3 +316,74 @@ export function unique<T>(...arrays: Array<readonly T[]>): T[] {
...
+ * const c = vCheck.checkedResult(a + b);
+ * const d = vCheck.checkedResult(c * c);
+ * const expectedResult = vCheck.allChecksPassed();
+ */
+export class ConstantOrOverrideValueChecker {
+  #allChecksPassed = true;
+  #floatLimits?: FloatLimits;
+  #quantizeFn: QuantizeFunc<number>;
+
+  constructor(private t: ShaderValidationTest, type: ScalarType) {
```
Comment (dneto0) [inline src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:344]: I don't know JS well enough but I presume this makes class fields `t` and `type`. Otherwise you couldn't access `t` to skip the test later.

**Review by dneto0 — COMMENTED** [review]
Looks pretty nice.

I'd like the isAmbiguousOverflow to be explicit about what it's supposed to do  with Infininties and naNs. (Well, Nans are the more interesting one, because the comparison at the end should take care of the infinity.)

And this would be improved with unit tests of isAmbiguousOverflow.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts [hunk 2]
```diff
@@ -312,3 +316,74 @@ export function unique<T>(...arrays: Array<readonly T[]>): T[] {
...
+  // Some overflow floating point values may fall into an abiguously rounded scenario, where they
+  // can either round up to Infinity or down to the maximum representable value. In these cases the
+  // test should be skipped, because it's valid for implementations to differ.
+  isAmbiguousOverflow(value: number): boolean {
+    if (!this.#floatLimits ||
+        (value <= this.#floatLimits.positive.max &&
+          value >= this.#floatLimits.negative.min)) {
+          return false;
+    }
+    return Math.abs(value) < Math.pow(2, this.#floatLimits.emax+1);
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:373]: I agree in about f64, since number is f64 we are already hitting issues with AF being an f64, if we add a concrete f64 (and presumably move AF to f128) there is going to need to be some major reworking in the CTS to support that.

File: src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts [hunk 3]
```diff
@@ -57,32 +45,22 @@ Validates that constant evaluation and override evaluation of ${builtin}() never
...
-    }
+    const ab = vCheck.checkedResult(a - b);
 
     // Only calculates the full length if the type is a vector. Otherwise abs(a-b) is used.
     if (kValidArgumentTypes[t.params.type].width > 1) {
-      const ab2 = quantizeFn(ab * ab);
-      const sqrLen = quantizeFn(ab2 * kValidArgumentTypes[t.params.type].width);
+      const ab2 = vCheck.checkedResult(ab * ab);
+      const sqrLen = vCheck.checkedResult(ab2 * kValidArgumentTypes[t.params.type].width);
       // Square root does not need to be calculated because it can never fail if
```
Comment (zoddicus) [inline src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts:62]: I don't think this comment is correct...
sqrt's accuracy is defined as 1/inverseSqrt(x)... which becomes a pain, because the inteval for inverseSqrt is 2 ULP, and division is 2.5 ULP plus not being define very close to 0 (https://www.w3.org/TR/WGSL/#floating-point-accuracy) So there should probably be a check that the value going into sqrt isn't near 0... and if it is, just skip, because it is likely to be a may fail case.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts [hunk 4]
```diff
@@ -312,3 +316,74 @@ export function unique<T>(...arrays: Array<readonly T[]>): T[] {
...
+  }
+
+  quantize(value: number): number {
+    return this.#quantizeFn(value);
+  }
+
+  // Some overflow floating point values may fall into an abiguously rounded scenario, where they
+  // can either round up to Infinity or down to the maximum representable value. In these cases the
+  // test should be skipped, because it's valid for implementations to differ.
+  isAmbiguousOverflow(value: number): boolean {
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:367]: From reading the code I think that for infinities it will always return false, because abs(+/-Infinit) is Infinity, which will never be less that 2^emax+1.

I think NaN just flows through and returns a false at the end, but I am not 100% sure about that.

PRobably worth being explict in the comment that values greater than 2^emax+1 will round to infinity and thus are not ambigious, i.e. will always fail validation

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:344]: Yeah, it makes them private fields. I had to look this up along with the # syntax, since I don't normally use them

**Review by zoddicus — CHANGES_REQUESTED** [review]
Overall I like it, other than the prexisting issue with sqrt handling that I think needs to be addressed, sorry for not catching that in an earlier review.

File: src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts (hunk 3 above)
Comment (toji) [reply src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts]: Thanks for pointing that out! Updated to add a near-zero check, implemented by testing to see if the value is between the min positive and max negative values for the type.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 1 above)
Comment (toji) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:346]: Yep, though it's a typescript feature, not a JavaScript one.

Happy to remove any syntax that feels out of place or confusing, I'm just used to using these in my own code.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (toji) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:370]: Added a bunch of comments and made the finite check explicit.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (toji) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts:384]: To clarify: Sounds like we don't need to do anything regarding f64 support now, but may need to in the future? Or are there changes that you'd like to see to how this is being handled today?

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts]: I am fine with leaving it as is, I don't find it particularly cryptic, was just a syntax I wasn't familar with.

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts]: No changes needed at this point (This is just mostly us waxing philospically about future evolutions of WGSL)

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 1 above)
Comment (toji) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts]: TypeScript allows for private attributes to also be declared as:

`private someValue: number;`

If that reads better.

**Review by zoddicus — APPROVED** [review]

File: src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/const_override_validation.ts]: I am fine with either way
